// www/index.js
module.exports = www;

function www() {}

www.prototype.start = function($happn, masterHappn, callback) {

  // $happn._mesh.datalayer.events.on('attach', function(ev) {
  //   console.log('ATTACH', JSON.stringify(ev, null, 2));
  // });

  // $happn._mesh.datalayer.events.on('detatch', function(ev) {
  //   console.log('DETATCH', JSON.stringify(ev, null, 2));
  // });

  this.master = masterHappn;

}

www.prototype.relayAction = function($happn, action, callback) {

  var component = action.component;
  var method = action.method;
  var params = action.params;

  $happn.log.$$DEBUG(
    'action component: \'%s\', method: \'%s\', params: \'%j\'',
    component,
    method,
    params
  );

  try {
    this.master.exchange[component][method](params, callback);
  }
  catch (e) {
    callback(e);
  }
}


// Master calls this to relay data to browser

www.prototype.update = function($happn, data, callback) {

  var marshals = [];

  for (var name in data.marshals) {

    var marshal = data.marshals[name];
    var minions = [];

    for (var minionName in marshal.minions) {
      minions.push({name: minionName, data: marshal.minions[minionName]});
    }

    marshals.push({name: name, minions: minions});
  }

  $happn.emit('update', {
    marshals: marshals
  });

  callback();

}
